File: grazioso_animal_shelter/frontend/src/api/animals.ts
import { apiClient } from "./client";

export interface Animal {
  id: number;
  animal_id: string;
  name: string | null;
  animal_type: string;
  breed: string;
  color: string | null;
  sex_upon_outcome: string | null;
  date_of_birth: string | null;
  outcome_type: string | null;
  outcome_subtype: string | null;
  outcome_datetime: string | null;
  age_upon_outcome_in_weeks: number | null;
  location_lat: number | null;
  location_long: number | null;
  archived_at: string | null;
}

export interface AnimalCreate {
  animal_id: string;
  name?: string | null;
  animal_type: string;
  breed: string;
  color?: string | null;
  sex_upon_outcome?: string | null;
  date_of_birth?: string | null;
  outcome_type?: string | null;
  outcome_subtype?: string | null;
  outcome_datetime?: string | null;
  age_upon_outcome_in_weeks?: number | null;
  location_lat?: number | null;
  location_long?: number | null;
}

export type AnimalUpdate = Partial<AnimalCreate>;

export interface AnimalPage {
  items: Animal[];
  total: number;
  page: number;
  page_size: number;
}

export interface AnimalSearchParams {
  q?: string;
  animalType?: string;
  page?: number;
  pageSize?: number;
  includeArchived?: boolean;
}

export const searchAnimals = (
  token: string,
  { q, animalType, page, pageSize, includeArchived }: AnimalSearchParams = {},
): Promise<AnimalPage> => {
  const params = new URLSearchParams();
  if (q) params.set("q", q);
  if (animalType) params.set("animal_type", animalType);
  if (page) params.set("page", String(page));
  if (pageSize) params.set("page_size", String(pageSize));
  if (includeArchived) params.set("include_archived", "true");

  const query = params.toString();
  return apiClient.get<AnimalPage>(`/animals${query ? `?${query}` : ""}`, token);
};

export const getAnimal = (token: string, animalPk: number): Promise<Animal> =>
  apiClient.get<Animal>(`/animals/${animalPk}`, token);

export const createAnimal = (token: string, payload: AnimalCreate): Promise<Animal> =>
  apiClient.post<Animal>("/animals", payload, token);

export const updateAnimal = (
  token: string,
  animalPk: number,
  payload: AnimalUpdate,
): Promise<Animal> => apiClient.patch<Animal>(`/animals/${animalPk}`, payload, token);

export const archiveAnimal = (token: string, animalPk: number): Promise<Animal> =>
  apiClient.post<Animal>(`/animals/${animalPk}/archive`, undefined, token);

export const unarchiveAnimal = (token: string, animalPk: number): Promise<Animal> =>
  apiClient.post<Animal>(`/animals/${animalPk}/unarchive`, undefined, token);

export interface BreedCount {
  breed: string;
  count: number;
}

export interface BreedSummary {
  items: BreedCount[];
  other_count: number;
  total_animals: number;
}

export interface BreedSummaryParams {
  q?: string;
  animalType?: string;
  limit?: number;
}

export const getBreedSummary = (
  token: string,
  { q, animalType, limit }: BreedSummaryParams = {},
): Promise<BreedSummary> => {
  const params = new URLSearchParams();
  if (q) params.set("q", q);
  if (animalType) params.set("animal_type", animalType);
  if (limit) params.set("limit", String(limit));

  const query = params.toString();
  return apiClient.get<BreedSummary>(`/animals/breed-summary${query ? `?${query}` : ""}`, token);
};

File: grazioso_animal_shelter/backend/app/api/routes/animals.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db, require_role
from app.schemas.animal import (
    AnimalCreate,
    AnimalOut,
    AnimalPage,
    AnimalUpdate,
    BreedCount,
    BreedSummary,
)
from app.services import animal_service

router = APIRouter(
    prefix="/animals",
    tags=["animals"],
    dependencies=[Depends(get_current_user)],
)

# Write operations are limited to shelter staff and admins.
manage_animals = Depends(require_role("staff", "admin"))


@router.get("", response_model=AnimalPage)
async def search_animals(
    db: Annotated[AsyncSession, Depends(get_db)],
    q: Annotated[str | None, Query(max_length=100)] = None,
    animal_type: Annotated[str | None, Query(max_length=50)] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    include_archived: Annotated[bool, Query()] = False,
) -> AnimalPage:
    animals, total = await animal_service.search_animals(
        db,
        q=q,
        animal_type=animal_type,
        page=page,
        page_size=page_size,
        include_archived=include_archived,
    )
    return AnimalPage(
        items=[AnimalOut.model_validate(a) for a in animals],
        total=total,
        page=page,
        page_size=page_size,
    )


# Registered before /{animal_pk} so the literal path wins route matching.
@router.get("/breed-summary", response_model=BreedSummary)
async def get_breed_summary(
    db: Annotated[AsyncSession, Depends(get_db)],
    q: Annotated[str | None, Query(max_length=100)] = None,
    animal_type: Annotated[str | None, Query(max_length=50)] = None,
    limit: Annotated[int, Query(ge=1, le=25)] = 10,
) -> BreedSummary:
    top, other_count, total_animals = await animal_service.breed_summary(
        db, q=q, animal_type=animal_type, limit=limit
    )
    return BreedSummary(
        items=[BreedCount(breed=breed, count=count) for breed, count in top],
        other_count=other_count,
        total_animals=total_animals,
    )


@router.get("/{animal_pk}", response_model=AnimalOut)
async def get_animal(animal_pk: int, db: Annotated[AsyncSession, Depends(get_db)]) -> AnimalOut:
    try:
        animal = await animal_service.get_animal(db, animal_pk)
    except animal_service.AnimalNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Animal not found"
        ) from exc
    return AnimalOut.model_validate(animal)


@router.post(
    "",
    response_model=AnimalOut,
    status_code=status.HTTP_201_CREATED,
    dependencies=[manage_animals],
)
async def create_animal(
    payload: AnimalCreate, db: Annotated[AsyncSession, Depends(get_db)]
) -> AnimalOut:
    try:
        animal = await animal_service.create_animal(db, payload.model_dump())
    except animal_service.UnknownLookupValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=f"Unknown value for {exc.field}"
        ) from exc
    return AnimalOut.model_validate(animal)


@router.patch("/{animal_pk}", response_model=AnimalOut, dependencies=[manage_animals])
async def update_animal(
    animal_pk: int, payload: AnimalUpdate, db: Annotated[AsyncSession, Depends(get_db)]
) -> AnimalOut:
    try:
        animal = await animal_service.update_animal(
            db, animal_pk, payload.model_dump(exclude_unset=True)
        )
    except animal_service.AnimalNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Animal not found"
        ) from exc
    except animal_service.UnknownLookupValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=f"Unknown value for {exc.field}"
        ) from exc
    return AnimalOut.model_validate(animal)


@router.post("/{animal_pk}/archive", response_model=AnimalOut, dependencies=[manage_animals])
async def archive_animal(
    animal_pk: int, db: Annotated[AsyncSession, Depends(get_db)]
) -> AnimalOut:
    try:
        animal = await animal_service.set_archived(db, animal_pk, archived=True)
    except animal_service.AnimalNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Animal not found"
        ) from exc
    return AnimalOut.model_validate(animal)


@router.post("/{animal_pk}/unarchive", response_model=AnimalOut, dependencies=[manage_animals])
async def unarchive_animal(
    animal_pk: int, db: Annotated[AsyncSession, Depends(get_db)]
) -> AnimalOut:
    try:
        animal = await animal_service.set_archived(db, animal_pk, archived=False)
    except animal_service.AnimalNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Animal not found"
        ) from exc
    return AnimalOut.model_validate(animal)
